test/javascripts/remote_test.coffee

```diff
@@ -67,6 +67,53 @@ describe 'Remote', ->
       assert.equal "/foo/bar", request.url
       assert.equal "POST", request.method
 
+  describe 'callbacks', ->
+
+    it 'will call options.fail() on HTTP failures', (done) ->
+      server = sinon.fakeServer.create()
+      server.respondWith("POST", "/foo/bar",
+            [422, { "Content-Type": "text/html" },
+             '<div id="foo" refresh="foo">Error occured</div>'])
+      remote = new TurboGraft.Remote
+        httpRequestType: "POST"
+        httpUrl: "/foo/bar"
+        refreshOnError: "foo"
+        fail: done
+      , @initiating_target
+      remote.submit()
+
+      server.respond()
+
+    it 'will call options.fail() on XHR failures', (done) ->
+      xhr = sinon.useFakeXMLHttpRequest()
+
+      remote = new TurboGraft.Remote
+        httpRequestType: "POST"
+        httpUrl: "/foo/bar"
+        refreshOnError: "foo"
+        fail: done
+      , @initiating_target
+
+      simulateError = ->
+        listener(target: remote.xhr) for listener in remote.xhr.eventListeners.error
+
+      setTimeout(simulateError, 0)
```

Comment:
To guarantee async. XHR error callbacks would never be called within the same tick.